File: 2025/src/py/day_one.py
# Copyright (C) 2025 Rémy Cases
# See LICENSE file for extended copyright information.
# This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

import ast
from dataclasses import dataclass
from pathlib import Path

@dataclass
class Rotation:
    increment: int = 0
    orientation: int = 0
    cycles: int = 0

MAX_DIAL = 100
LEFT = 1
RIGHT = -1

def parseRotation(line: str) -> Rotation:
    r = Rotation()
    if line[0] == 'L':
        r.orientation = LEFT
    elif line[0] == 'R':
        r.orientation = RIGHT
    else:
        raise ValueError(f"Invalid line to parse: {line}.")

    val = ast.literal_eval(line[1:])

    r.increment = val % MAX_DIAL
    r.cycles = val // MAX_DIAL
    return r

def readFileAndCompute(filename: Path, option: int = 1) -> None:
    dial: int = 50
    dial_at_zero: int = 0

    with open(filename, 'r') as file:
        for line in file:
            rotation = parseRotation(line)

            if option == 1:
                if rotation.increment == 0:
                    continue

                if rotation.orientation == LEFT:
                    dial = (dial + MAX_DIAL - rotation.increment) % MAX_DIAL
                elif rotation.orientation == RIGHT:
                    dial = (dial + rotation.increment) % MAX_DIAL
                if dial == 0:
                    dial_at_zero += 1

            elif option == 2:
                dial_at_zero += rotation.cycles
                if rotation.increment == 0:
                    continue
    
                if rotation.orientation == LEFT:
                    if dial <= rotation.increment and dial != 0:
                        dial_at_zero += 1
                    dial = (dial + MAX_DIAL - rotation.increment) % MAX_DIAL
                elif rotation.orientation == RIGHT:
                    if dial + rotation.increment >= MAX_DIAL:
                        dial_at_zero += 1
                    dial = (dial + rotation.increment) % MAX_DIAL
            else:
                print("Invalid part")
                return

    print(f"DIAL AT ZERO: {dial_at_zero}")

def main(part: int) -> None:
    file = Path("2025/data/input_day_one")
    filename = Path.cwd() / file
    readFileAndCompute(filename, part)

File: 2025/src/lua/dayOne.lua
-- Copyright (C) 2025 Rémy Cases
-- See LICENSE file for extended copyright information.
-- This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

local DialMax = 100
local Left = 1
local Right = -1

local function parseRotation(line)
    local o = line:byte(1)
    local rotation = {["orientation"]= 0, ["increment"]= 0, ["cycles"] = 0}
    if o == string.byte('L') then
        rotation.orientation = Left
    elseif o == string.byte('R') then
        rotation.orientation = Right
    else
        error(string.format("Error parsing %s", line))
    end

    local val = tonumber(string.sub(line,2))

    if not val then
        error(string.format("Error parsing %s", line))
    end

    rotation.increment = val % DialMax
    rotation.cycles = val // DialMax
    return rotation
end

local function readFileAndCompute(filename, option)
    local dial_at_zero = 0
    local dial = 50

    local file = io.open(filename,"rb")
    if not file then
        return nil
    end
    file:close()

    for line in io.lines(filename) do
        local rotation = parseRotation(line)

        if option == '1' then
            if rotation.increment == 0 then
                goto continue
            end

            if rotation.orientation == Left then
                dial = (dial + DialMax - rotation.increment) % DialMax
            elseif rotation.orientation == Right then
                dial = (dial + rotation.increment) % DialMax
            end

            if dial == 0 then
                dial_at_zero = dial_at_zero + 1
            end

        elseif option == '2' then
            dial_at_zero = dial_at_zero + rotation.cycles
            if rotation.increment == 0 then
                goto continue
            end

            if rotation.orientation == Left then
                if dial <= rotation.increment and dial ~= 0 then
                    dial_at_zero = dial_at_zero + 1
                end
                dial = (dial + DialMax - rotation.increment) % DialMax
            elseif rotation.orientation == Right then
                if dial + rotation.increment >= DialMax then
                    dial_at_zero = dial_at_zero + 1
                end
                dial = (dial + rotation.increment) % DialMax
            end
        else
            print("Invalid part")
            return
        end

        ::continue::
    end
    print(string.format("DIAL AT ZERO: %d", dial_at_zero))
end

-- table of functions accessibles through require
local M = { }

local function main(part)
    local filename = "2025/data/input_day_one"
    readFileAndCompute(filename, part)
end
M.main = main

return M